multimodal/src/autogluon/multimodal/utils/onnx.py

```diff
@@ -7,99 +7,100 @@
-                    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
-                    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
-                    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
-                ],
-            ),
-        }
-    elif pipeline == MULTICLASS:
-        valid_input = ["timm_image_image", "timm_image_image_valid_num"]
-        dynamic_axes = {
-            "timm_image_image": {
-                0: "batch_size",
-            },
-            "timm_image_image_valid_num": {},
-        }
-        default_onnx_path = "./model.onnx"
-        default_batch = None
-    else:
-        raise ValueError(f"ONNX export is not supported in current pipeline {pipeline}")
-
-    return valid_input, dynamic_axes, default_onnx_path, default_batch
+
+    def __init__(self, model, providers: Optional[Union[dict, List[str]]] = None):
+        """
+        Parameters
+        ----------
+        model : onnx.ModelProto
+            The onnx model that need to be executed in onnxruntime.
+        providers : dict or str, default=None
+            A list of execution providers for model prediction in onnxruntime.
+        """
+        import onnxruntime as ort
+
+        if providers == None:
+            providers = [
+                (
+                    "TensorrtExecutionProvider",
+                    {
+                        "device_id": 0,
+                        "trt_max_workspace_size": 2147483648,
+                        "trt_fp16_enable": True,
+                    },
+                ),
+                (
+                    "CUDAExecutionProvider",
+                    {
+                        "device_id": 0,
+                        "arena_extend_strategy": "kNextPowerOfTwo",
+                        "gpu_mem_limit": 2 * 1024 * 1024 * 1024,
+                        "cudnn_conv_algo_search": "EXHAUSTIVE",
+                        "do_copy_in_default_stream": True,
+                    },
+                ),
+                ("CPUExecutionProvider", {}),
+            ]
+        self.sess = ort.InferenceSession(model.SerializeToString(), providers=providers)
+        inputs = self.sess.get_inputs()
+        outputs = self.sess.get_outputs()
+        self.input_names = [i.name for i in inputs]
+        self.output_names = [i.name for i in outputs]
+
+    def __call__(self, *args):
+        """
+        Make the module callable like torch.nn.Module, while runnning forward pass with onnxruntime.
+
+        Parameters
+        ----------
+        args : list of torch.Tensor
+            A list of torch.Tensor that are inputs of the model.
+
+        Returns
+        -------
+        onnx_outputs : list of torch.Tensor
+            A list of torch.Tensor that are outputs of the model.
+        """
+        import torch
+
+        input_dict = {k: args[i].cpu().numpy() for i, k in enumerate(self.input_names)}
+        onnx_outputs = self.sess.run(self.output_names, input_dict)
+        onnx_outputs = onnx_outputs[:3]
```

Comment:
Why do we need only the first 3 model outputs? Don't we need them all?

Reply:
This comes with the undetermined size of output dict for a fusion_mlp model, which contains a tuple of `(features, logits, multimodal_logits)`. The variable `multimodal_logits` contains the logit output from all modalities.

> Don't we need them all?

Good question. In short, `multimodal_logits` are only used for computing loss, not used for inference.

Specifically, `get_output_dict()` would take outputs from `forward()`, but in onnxruntime outputs are flatten, which means the list of tensors in `multimodal_logits` would be merged into other outputs (e.g. features, logits). We kind of lost the information about where did the extra tensors come from.